libbeat/testing/integration/integration.go

```diff
@@ -0,0 +1,336 @@
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package integration
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os/exec"
+	"regexp"
+	"sync"
+	"testing"
+)
+
+const (
+	expectErrMsg = "cannot set expectations once the test started"
+)
+
+// BeatTest describes all operations involved
+// in integration testing of a Beat
+type BeatTest interface {
+	// Start the integration test
+	//
+	// The test runs until all the expectations are met (unless `ExpectStop` is used) or context was canceled or the Beat exits on its own.
+	Start(context.Context) BeatTest
+
+	// Wait until the test is over.
+	//
+	// `PrintOutput` might be helpful for debugging after calling this function.
+	Wait()
+
+	// ExpectStart sets an expectation that the Beat will report that it started.
+	ExpectStart() BeatTest
+
+	// ExpectStop sets an expectation that the Beat will exit by itself.
+	// The process exit code will be checked against the given value.
+	//
+	// User controls the timeout by passing the context in `Start`.
+	//
+	// All the output expectations would still work as usual, however,
+	// satisfying all expectations would not stop the Beat.
+	ExpectStop(exitCode int) BeatTest
+
+	// ExpectOutput registers an output watch for the given substrings.
+	//
+	// Every future output line produced by the Beat will be checked
+	// if it contains one of the given strings.
+	//
+	// If given multiple strings, they get checked in order:
+	// The first substring must be found first, then second, etc.
+	//
+	// For `AND` behavior use this function multiple times.
+	//
+	// This function should be used before `Start` because it's
+	// inspecting only the new output lines.
+	ExpectOutput(...string) BeatTest
+
+	// ExpectOutputRegex registers an output watch for the given regular expression..
+	//
+	// Every future output line produced by the Beat will be matched
+	// against the given regular expression.
+	//
+	// If given multiple expressions, they get checked in order.
+	// The first expression must match first, then second, etc.
+	//
+	// For `AND` behavior use this function multiple times.
+	//
+	// This function should be used before `Start` because it's
+	// inspecting only new outputs.
+	ExpectOutputRegex(...*regexp.Regexp) BeatTest
```

Comment:
[NIT]
You could say something like `Same as ExpectOutput, but using a regex instead` and avoid the text duplication

Reply:
When you use an IDE you cannot see documentation of other functions in autocomplete.